#Deep Learning With Computer Vision And Advanced NLP (DL_CV_NLP)

$$ Revision Notes $$
$$ A-Note-by-**Bappy Ahmed** $$

# Observing results before and after applying Batch Normalization:

## Before applying Batch Normalization:

In [ ]:
# Importing necessary libraries
import os
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import time
plt.style.use("fivethirtyeight")
%load_ext tensorboard

The tensorboard extension is already loaded. To reload it, use:
  %reload_ext tensorboard


In [ ]:
# Loading the data of fashion_mnist
(X_train_full, y_train_full), (X_test, y_test) = tf.keras.datasets.fashion_mnist.load_data()
X_train_full = X_train_full / 255.0
X_test = X_test / 255.0
X_valid, X_train = X_train_full[:5000], X_train_full[5000:]
y_valid, y_train = y_train_full[:5000], y_train_full[5000:]

In [ ]:
# Creating layer of model
tf.random.set_seed(42) #For getting similar output (optional)
np.random.seed(42) #For getting similar output (optional)

LAYERS = [ tf.keras.layers.Flatten(input_shape=[28, 28]),
    tf.keras.layers.Dense(300, kernel_initializer="he_normal"),
    tf.keras.layers.LeakyReLU(),
    tf.keras.layers.Dense(100, kernel_initializer="he_normal"),
    tf.keras.layers.LeakyReLU(),
    tf.keras.layers.Dense(10, activation="softmax")]


model = tf.keras.models.Sequential(LAYERS)

In [ ]:
# Compiling the model
model.compile(loss="sparse_categorical_crossentropy",
              optimizer=tf.keras.optimizers.SGD(lr=1e-3),
              metrics=["accuracy"])

/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/optimizer_v2/optimizer_v2.py:375: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  "The `lr` argument is deprecated, use `learning_rate` instead.")


In [ ]:
model.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_2 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_6 (Dense)              (None, 300)               235500    
_________________________________________________________________
leaky_re_lu_2 (LeakyReLU)    (None, 300)               0         
_________________________________________________________________
dense_7 (Dense)              (None, 100)               30100     
_________________________________________________________________
leaky_re_lu_3 (LeakyReLU)    (None, 100)               0         
_________________________________________________________________
dense_8 (Dense)              (None, 10)                1010      
Total params: 266,610
Trainable params: 266,610
Non-trainable params: 0
________________________________________________

In [ ]:
# now training & calculating the training time.

# starting time
start = time.time()


history = model.fit(X_train, y_train, epochs=10,
                    validation_data=(X_valid, y_valid), verbose=2)

#ending time
end = time.time()

# total time taken
print(f"Runtime of the program is {end - start}")

Epoch 1/10
1719/1719 - 3s - loss: 1.2819 - accuracy: 0.6229 - val_loss: 0.8886 - val_accuracy: 0.7160
Epoch 2/10
1719/1719 - 3s - loss: 0.7955 - accuracy: 0.7362 - val_loss: 0.7130 - val_accuracy: 0.7658
Epoch 3/10
1719/1719 - 3s - loss: 0.6816 - accuracy: 0.7720 - val_loss: 0.6427 - val_accuracy: 0.7898
Epoch 4/10
1719/1719 - 3s - loss: 0.6217 - accuracy: 0.7944 - val_loss: 0.5900 - val_accuracy: 0.8064
Epoch 5/10
1719/1719 - 3s - loss: 0.5832 - accuracy: 0.8074 - val_loss: 0.5582 - val_accuracy: 0.8202
Epoch 6/10
1719/1719 - 3s - loss: 0.5553 - accuracy: 0.8156 - val_loss: 0.5350 - val_accuracy: 0.8236
Epoch 7/10
1719/1719 - 3s - loss: 0.5338 - accuracy: 0.8225 - val_loss: 0.5157 - val_accuracy: 0.8304
Epoch 8/10
1719/1719 - 3s - loss: 0.5173 - accuracy: 0.8272 - val_loss: 0.5079 - val_accuracy: 0.8284
Epoch 9/10
1719/1719 - 3s - loss: 0.5040 - accuracy: 0.8289 - val_loss: 0.4895 - val_accuracy: 0.8388
Epoch 10/10
1719/1719 - 3s - loss: 0.4924 - accuracy: 0.8321 - val_loss: 0.4817 - 

## Conclution:
- Runtime of the program is 26.92 sec
- accuracy: 0.8321

# After applying Batch Normalization:

In [ ]:
# delete the previous model
del model

In [ ]:
# Defing new model with batch normalization
LAYERS_BN = [
    tf.keras.layers.Flatten(input_shape=[28, 28]),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(300, activation="relu"),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(100, activation="relu"),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(10, activation="softmax")
]

model = tf.keras.models.Sequential(LAYERS_BN)

In [ ]:
model.summary()

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_3 (Flatten)          (None, 784)               0         
_________________________________________________________________
batch_normalization_3 (Batch (None, 784)               3136      
_________________________________________________________________
dense_9 (Dense)              (None, 300)               235500    
_________________________________________________________________
batch_normalization_4 (Batch (None, 300)               1200      
_________________________________________________________________
dense_10 (Dense)             (None, 100)               30100     
_________________________________________________________________
batch_normalization_5 (Batch (None, 100)               400       
_________________________________________________________________
dense_11 (Dense)             (None, 10)               

In [ ]:
bn1 = model.layers[1]

In [ ]:
for variable in bn1.variables:
  print(variable.name, variable.trainable)

batch_normalization_3/gamma:0 True
batch_normalization_3/beta:0 True
batch_normalization_3/moving_mean:0 False
batch_normalization_3/moving_variance:0 False


In [ ]:
model.compile(loss="sparse_categorical_crossentropy",
              optimizer=tf.keras.optimizers.SGD(lr=1e-3),
              metrics=["accuracy"])

/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/optimizer_v2/optimizer_v2.py:375: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  "The `lr` argument is deprecated, use `learning_rate` instead.")


In [ ]:
# now training & calculating the training time.

# starting time
start = time.time()


history = model.fit(X_train, y_train, epochs=10,
                    validation_data=(X_valid, y_valid), verbose=2)

#ending time
end = time.time()

# total time taken
print(f"Runtime of the program is {end - start}")

Epoch 1/10
1719/1719 - 5s - loss: 0.8293 - accuracy: 0.7221 - val_loss: 0.5539 - val_accuracy: 0.8160
Epoch 2/10
1719/1719 - 4s - loss: 0.5703 - accuracy: 0.8035 - val_loss: 0.4792 - val_accuracy: 0.8374
Epoch 3/10
1719/1719 - 4s - loss: 0.5161 - accuracy: 0.8214 - val_loss: 0.4424 - val_accuracy: 0.8488
Epoch 4/10
1719/1719 - 4s - loss: 0.4789 - accuracy: 0.8316 - val_loss: 0.4212 - val_accuracy: 0.8560
Epoch 5/10
1719/1719 - 4s - loss: 0.4548 - accuracy: 0.8406 - val_loss: 0.4051 - val_accuracy: 0.8612
Epoch 6/10
1719/1719 - 4s - loss: 0.4386 - accuracy: 0.8445 - val_loss: 0.3932 - val_accuracy: 0.8632
Epoch 7/10
1719/1719 - 4s - loss: 0.4254 - accuracy: 0.8505 - val_loss: 0.3829 - val_accuracy: 0.8646
Epoch 8/10
1719/1719 - 4s - loss: 0.4123 - accuracy: 0.8538 - val_loss: 0.3760 - val_accuracy: 0.8666
Epoch 9/10
1719/1719 - 4s - loss: 0.4027 - accuracy: 0.8579 - val_loss: 0.3690 - val_accuracy: 0.8674
Epoch 10/10
1719/1719 - 4s - loss: 0.3925 - accuracy: 0.8616 - val_loss: 0.3630 - 

## Conclution:
- Runtime of the program is 44.89 sec
- accuracy: 0.8616

# Comparison:

## Before applying
- Runtime of the program is 26.92 sec
- accuracy: 0.8321

## After applying
- Runtime of the program is 44.89 sec
- accuracy: 0.8616

**Note**: As we can see although it has been taken a little long time due to the some extra parameters added but it has been improved the accurary score after applying BN. In big problem it performs more.